app/jobs/catalog_job.rb

```diff
@@ -34,6 +34,8 @@ def update_registry!(catalog)
 
     # At this point, the remaining items in the "repos" array are repos that
     # exist in the DB but not in the catalog. Remove all of them.
-    Repository.where(id: dangling_repos).delete_all
+    repos = Repository.where(id: dangling_repos)
+    repos.find_each { |r| r.tags.find_each(&:delete_and_update!) }
```

Comment:
you need n+1 this way. Maybe `Tag.where(repository_id: dangling_repos).find_each(&:delete_and_update!)` is better.